[Turbopack Nextjs 16]: Pino - Worker thread cannot find module 'real-require'

### Link to the code that reproduces this issue

https://github.com/CHC383/nextjs-turbopack-pino-reproduction

### To Reproduce

1. Deploy the code to Vercel
1. Visit `<url>/api/log`
1. Check the log of the Vercel deployment

### Current vs. Expected behavior

**Current behavior**

With Turbopack in NextJs v16, Pino worker thread throws the following error:

```
Uncaught Exception: Error: Cannot find module 'real-require'
Require stack:
- /var/task/node_modules/.pnpm/thread-stream@3.1.0/node_modules/thread-stream/lib/worker.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1365:15)
    at defaultResolveImpl (node:internal/modules/cjs/loader:1021:19)
    at resolveForCJSWithHooks (node:internal/modules/cjs/loader:1026:22)
    at Module._load (node:internal/modules/cjs/loader:1175:37)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:235:24)
    at Module.require (node:internal/modules/cjs/loader:1445:12)
    at require (node:internal/modules/helpers:135:16)
    at Object.<anonymous> (/var/task/node_modules/.pnpm/thread-stream@3.1.0/node_modules/thread-stream/lib/worker.js:3:37)
    at Module._compile (node:internal/modules/cjs/loader:1688:14)
```

**Expected behavior**

Pino worker thread works without any exception, which is the behavior of using
WebPack in NextJs v15.


### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 25.0.0: Wed Sep 17 21:41:45 PDT 2025; root:xnu-12377.1.9~141/RELEASE_ARM64_T6000
  Available memory (MB): 16384
  Available CPU cores: 10
Binaries:
  Node: 22.20.0
  npm: 10.9.3
  Yarn: 1.22.22
  pnpm: 10.18.1
Relevant Packages:
  next: 16.0.0-beta.0 // Latest available version is detected (16.0.0-beta.0).
  eslint-config-next: N/A
  react: 19.2.0
  react-dom: 19.2.0
  typescript: 5.9.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Turbopack

### Which stage(s) are affected? (Select all that apply)

Vercel (Deployed)

### Additional context

The transport target `pino-pretty` in the reproduction repository is just for demo purpose, the same issue happens on any of the targets.

Pino uses [thread-stream](https://github.com/pinojs/thread-stream) to send the logs to worker thread when logging asynchronously, which uses [real-require](https://github.com/pinojs/real-require) to work with webpack import. Looks like Turbopack doesn't bundle`real-require` properly.

<sub>[PACK-5694](https://linear.app/vercel/issue/PACK-5694/turbopack-nextjs-16-pino-worker-thread-cannot-find-module-real-require)</sub>

--- a/turbopack/crates/turbopack-core/src/reference_type.rs
+++ b/turbopack/crates/turbopack-core/src/reference_type.rs
@@ -290,6 +290,7 @@ pub enum WorkerReferenceSubType {
     WebWorker,
     SharedWorker,
     ServiceWorker,
+    NodeWorker,
     Custom(u8),
     Undefined,
 }

--- a/turbopack/crates/turbopack-ecmascript/src/analyzer/mod.rs
+++ b/turbopack/crates/turbopack-ecmascript/src/analyzer/mod.rs
@@ -1788,7 +1788,11 @@ impl JsValue {
                         "module",
                         "The Node.js `module` module: https://nodejs.org/api/module.html",
                     ),
-                    WellKnownObjectKind::ChildProcess | WellKnownObjectKind::ChildProcessDefault => (
+                    WellKnownObjectKind::WorkerThreadsModule | WellKnownObjectKind::WorkerThreadsModuleDefault => (
+                        "worker_threads",
+                        "The Node.js `worker_threads` module: https://nodejs.org/api/worker_threads.html",
+                    ),
+                    WellKnownObjectKind::ChildProcessModule | WellKnownObjectKind::ChildProcessModuleDefault => (
                         "child_process",
                         "The Node.js child_process module: https://nodejs.org/api/child_process.html",
                     ),
@@ -1796,7 +1800,7 @@ impl JsValue {
                         "os",
                         "The Node.js os module: https://nodejs.org/api/os.html",
                     ),
-                    WellKnownObjectKind::NodeProcess => (
+                    WellKnownObjectKind::NodeProcessModule => (
                         "process",
                         "The Node.js process module: https://nodejs.org/api/process.html",
                     ),
@@ -1953,6 +1957,10 @@ impl JsValue {
                     WellKnownFunctionKind::NodeProtobufLoad => (
                       "load/loadSync".to_string(),
                       "require('@grpc/proto-loader').load(filepath, { includeDirs: [root] }) https://github.com/grpc/grpc-node"
+                    ),
+                    WellKnownFunctionKind::NodeWorkerConstructor => (
+                      "Worker".to_string(),
+                      "The Node.js worker_threads Worker constructor: https://nodejs.org/api/worker_threads.html#worker_threads_class_worker"
                     ),
                     WellKnownFunctionKind::WorkerConstructor => (
                       "Worker".to_string(),
@@ -3468,11 +3476,13 @@ pub enum WellKnownObjectKind {
     ModuleModuleDefault,
     UrlModule,
     UrlModuleDefault,
-    ChildProcess,
-    ChildProcessDefault,
+    WorkerThreadsModule,
+    WorkerThreadsModuleDefault,
+    ChildProcessModule,
+    ChildProcessModuleDefault,
     OsModule,
     OsModuleDefault,
-    NodeProcess,
+    NodeProcessModule,
     NodeProcessArgv,
     NodeProcessEnv,
     NodePreGyp,
@@ -3492,9 +3502,10 @@ impl WellKnownObjectKind {
             Self::PathModule => Some(&["path"]),
             Self::FsModule => Some(&["fs"]),
             Self::UrlModule => Some(&["url"]),
-            Self::ChildProcess => Some(&["child_process"]),
+            Self::ChildProcessModule => Some(&["child_process"]),
             Self::OsModule => Some(&["os"]),
-            Self::NodeProcess => Some(&["process"]),
+            Self::WorkerThreadsModule => Some(&["worker_threads"]),
+            Self::NodeProcessModule => Some(&["process"]),
             Self::NodeProcessArgv => Some(&["process", "argv"]),
             Self::NodeProcessEnv => Some(&["process", "env"]),
             Self::NodeBuffer => Some(&["Buffer"]),
@@ -3623,6 +3634,8 @@ pub enum WellKnownFunctionKind {
     NodeResolveFrom,
     NodeProtobufLoad,
     WorkerConstructor,
+    // The worker_threads Worker class
+    NodeWorkerConstructor,
     URLConstructor,
 }
 
@@ -3797,7 +3810,7 @@ pub mod test_utils {
                 ),
                 "define" => JsValue::WellKnownFunction(WellKnownFunctionKind::Define),
                 "URL" => JsValue::WellKnownFunction(WellKnownFunctionKind::URLConstructor),
-                "process" => JsValue::WellKnownObject(WellKnownObjectKind::NodeProcess),
+                "process" => JsValue::WellKnownObject(WellKnownObjectKind::NodeProcessModule),
                 "Object" => JsValue::WellKnownObject(WellKnownObjectKind::GlobalObject),
                 "Buffer" => JsValue::WellKnownObject(WellKnownObjectKind::NodeBuffer),
                 _ => v.into_unknown(true, "unknown global"),

--- a/turbopack/crates/turbopack-ecmascript/src/analyzer/well_known.rs
+++ b/turbopack/crates/turbopack-ecmascript/src/analyzer/well_known.rs
@@ -609,13 +609,18 @@ async fn well_known_object_member(
         WellKnownObjectKind::UrlModule | WellKnownObjectKind::UrlModuleDefault => {
             url_module_member(kind, prop)
         }
-        WellKnownObjectKind::ChildProcess | WellKnownObjectKind::ChildProcessDefault => {
-            child_process_module_member(kind, prop)
+        WellKnownObjectKind::WorkerThreadsModule
+        | WellKnownObjectKind::WorkerThreadsModuleDefault => {
+            worker_threads_module_member(kind, prop)
         }
+        WellKnownObjectKind::ChildProcessModule
+        | WellKnownObjectKind::ChildProcessModuleDefault => child_process_module_member(kind, prop),
         WellKnownObjectKind::OsModule | WellKnownObjectKind::OsModuleDefault => {
             os_module_member(kind, prop)
         }
-        WellKnownObjectKind::NodeProcess => node_process_member(prop, compile_time_info).await?,
+        WellKnownObjectKind::NodeProcessModule => {
+            node_process_member(prop, compile_time_info).await?
+        }
         WellKnownObjectKind::NodePreGyp => node_pre_gyp(prop),
         WellKnownObjectKind::NodeExpressApp => express(prop),
         WellKnownObjectKind::NodeProtobufLoader => protobuf_loader(prop),
@@ -762,7 +767,7 @@ fn url_module_member(kind: WellKnownObjectKind, prop: JsValue) -> JsValue {
         (.., Some("pathToFileURL")) => {
             JsValue::WellKnownFunction(WellKnownFunctionKind::PathToFileUrl)
         }
-        (WellKnownObjectKind::UrlModuleDefault, Some("default")) => {
+        (WellKnownObjectKind::UrlModule, Some("default")) => {
             JsValue::WellKnownObject(WellKnownObjectKind::UrlModuleDefault)
         }
         _ => JsValue::unknown(
@@ -776,6 +781,27 @@ fn url_module_member(kind: WellKnownObjectKind, prop: JsValue) -> JsValue {
     }
 }
 
+fn worker_threads_module_member(kind: WellKnownObjectKind, prop: JsValue) -> JsValue {
+    match (kind, prop.as_str()) {
+        (.., Some("Worker")) => {
+            JsValue::WellKnownFunction(WellKnownFunctionKind::NodeWorkerConstructor)
+        }
+        (WellKnownObjectKind::WorkerThreadsModule, Some("default")) => {
+            JsValue::WellKnownObject(WellKnownObjectKind::WorkerThreadsModuleDefault)
+        }
+        _ => JsValue::unknown(
+            JsValue::member(
+                Box::new(JsValue::WellKnownObject(
+                    WellKnownObjectKind::WorkerThreadsModule,
+                )),
+                Box::new(prop),
+            ),
+            true,
+            "unsupported property on Node.js worker_threads module",
+        ),
+    }
+}
+
 fn child_process_module_member(kind: WellKnownObjectKind, prop: JsValue) -> JsValue {
     let prop_str = prop.as_str();
     match (kind, prop_str) {
@@ -785,13 +811,15 @@ fn child_process_module_member(kind: WellKnownObjectKind, prop: JsValue) -> JsVa
             ))
         }
         (.., Some("fork")) => JsValue::WellKnownFunction(WellKnownFunctionKind::ChildProcessFork),
-        (WellKnownObjectKind::ChildProcess, Some("default")) => {
-            JsValue::WellKnownObject(WellKnownObjectKind::ChildProcessDefault)
+        (WellKnownObjectKind::ChildProcessModule, Some("default")) => {
+            JsValue::WellKnownObject(WellKnownObjectKind::ChildProcessModuleDefault)
         }
 
         _ => JsValue::unknown(
             JsValue::member(
-                Box::new(JsValue::WellKnownObject(WellKnownObjectKind::ChildProcess)),
+                Box::new(JsValue::WellKnownObject(
+                    WellKnownObjectKind::ChildProcessModule,
+                )),
                 Box::new(prop),
             ),
             true,
@@ -843,7 +871,9 @@ async fn node_process_member(
         Some("env") => JsValue::WellKnownObject(WellKnownObjectKind::NodeProcessEnv),
         _ => JsValue::unknown(
             JsValue::member(
-                Box::new(JsValue::WellKnownObject(WellKnownObjectKind::NodeProcess)),
+                Box::new(JsValue::WellKnownObject(
+                    WellKnownObjectKind::NodeProcessModule,
+                )),
                 Box::new(prop),
             ),
             true,

--- a/turbopack/crates/turbopack-ecmascript/src/parse.rs
+++ b/turbopack/crates/turbopack-ecmascript/src/parse.rs
@@ -176,15 +176,9 @@ pub async fn parse(
         ty = display(&ty)
     );
 
-    match parse_internal(
-        source,
-        ty,
-        transforms,
-        is_external_tracing && matches!(ty, EcmascriptModuleAssetType::EcmascriptExtensionless),
-        inline_helpers,
-    )
-    .instrument(span)
-    .await
+    match parse_internal(source, ty, transforms, is_external_tracing, inline_helpers)
+        .instrument(span)
+        .await
     {
         Ok(result) => Ok(result),
         Err(error) => Err(error.context(format!(

--- a/turbopack/crates/turbopack-ecmascript/src/references/mod.rs
+++ b/turbopack/crates/turbopack-ecmascript/src/references/mod.rs
@@ -151,7 +151,7 @@ use crate::{
         },
         ident::IdentReplacement,
         member::MemberReplacement,
-        node::PackageJsonReference,
+        node::{FilePathModuleReference, PackageJsonReference},
         require_context::{RequireContextAssetReference, RequireContextMap},
         type_issue::SpecifiedModuleTypeIssue,
     },
@@ -1646,8 +1646,27 @@ async fn handle_call<G: Fn(Vec<Effect>) + Send + Sync>(
             .await
     };
 
-    let make_issue_source =
+    let get_traced_project_dir = async || -> Result<FileSystemPath> {
+        // readFileSync("./foo") should always be relative to the project root, but this is
+        // dangerous inside of node_modules as it can cause a lot of false positives in the tracing,
+        // if some package does `path.join(dynamic)`, it would include everything from the project
+        // root as well.
+        //
+        // Also, when there's no cwd set (i.e. in a tracing-specific module context, as we shouldn't
+        // assume a `process.cwd()` for all of node_modules), fallback to the source file directory.
+        // This still allows relative file accesses, just not from the project root.
+        if allow_project_root_tracing
+            && let Some(cwd) = compile_time_info.environment().cwd().owned().await?
+        {
+            Ok(cwd)
+        } else {
+            Ok(source.ident().path().await?.parent())
+        }
+    };
+
+    let get_issue_source =
         || IssueSource::from_swc_offsets(source, span.lo.to_u32(), span.hi.to_u32());
+
     if new {
         match func {
             JsValue::WellKnownFunction(WellKnownFunctionKind::URLConstructor) => {
@@ -1729,10 +1748,10 @@ async fn handle_call<G: Fn(Vec<Effect>) + Send + Sync>(
             JsValue::WellKnownFunction(WellKnownFunctionKind::NodeWorkerConstructor)
                 if analysis.analyze_mode.is_tracing() =>
             {
-                // Only for tracing, not for bundling
+                // Only for tracing, not for bundling (yet?)
                 let args = linked_args(args).await?;
-                if let Some(filename) = args.first() {
-                    let pat = js_value_to_pattern(filename);
+                if !args.is_empty() {
+                    let pat = js_value_to_pattern(&args[0]);
                     if !pat.has_constant_parts() {
                         let (args, hints) = explain_args(&args);
                         handler.span_warn_with_code(
@@ -1746,25 +1765,34 @@ async fn handle_call<G: Fn(Vec<Effect>) + Send + Sync>(
                             return Ok(());
                         }
                     }
-
-                    analysis.add_reference_code_gen(
-                        WorkerAssetReference::new(
-                            origin,
-                            Request::parse(pat).to_resolved().await?,
-                            issue_source(source, span),
-                            in_try,
-                        ),
-                        ast_path.to_vec().into(),
+                    analysis.add_reference(
+                        FilePathModuleReference::new(
+                            origin.asset_context(),
+                            get_traced_project_dir().await?,
+                            Pattern::new(pat),
+                            collect_affecting_sources,
+                            get_issue_source(),
+                        )
+                        .to_resolved()
+                        .await?,
                     );
-
                     return Ok(());
                 }
-                // Ignore (e.g. dynamic parameter or string literal), just as Webpack does
+                let (args, hints) = explain_args(&args);
+                handler.span_warn_with_code(
+                    span,
+                    &format!("new Worker({args}) is not statically analyze-able{hints}",),
+                    DiagnosticId::Error(
+                        errors::failed_to_analyze::ecmascript::FS_METHOD.to_string(),
+                    ),
+                );
+                // Ignore (e.g. dynamic parameter or string literal)
                 return Ok(());
             }
             _ => {}
         }
 
+        // linked_args wasn't called, so manually add the closure effects
         for arg in args {
             if let EffectArg::Closure(_, block) = arg {
                 add_effects(block.effects);
@@ -1772,24 +1800,6 @@ async fn handle_call<G: Fn(Vec<Effect>) + Send + Sync>(
         }
         return Ok(());
     }
-
-    let get_traced_project_dir = async || -> Result<FileSystemPath> {
-        // readFileSync("./foo") should always be relative to the project root, but this is
-        // dangerous inside of node_modules as it can cause a lot of false positives in the tracing,
-        // if some package does `path.join(dynamic)`, it would include everything from the project
-        // root as well.
-        //
-        // Also, when there's no cwd set (i.e. in a tracing-specific module context, as we shouldn't
-        // assume a `process.cwd()` for all of node_modules), fallback to the source file directory.
-        // This still allows relative file accesses, just not from the project root.
-        if allow_project_root_tracing
-            && let Some(cwd) = compile_time_info.environment().cwd().owned().await?
-        {
-            Ok(cwd)
-        } else {
-            Ok(source.ident().path().await?.parent())
-        }
-    };
 
     match func {
         JsValue::Alternatives {
@@ -2024,7 +2034,7 @@ async fn handle_call<G: Fn(Vec<Effect>) + Send + Sync>(
                         get_traced_project_dir().await?,
                         Pattern::new(pat),
                         collect_affecting_sources,
-                        make_issue_source(),
+                        get_issue_source(),
                     )
                     .to_resolved()
                     .await?,
@@ -2077,7 +2087,7 @@ async fn handle_call<G: Fn(Vec<Effect>) + Send + Sync>(
                 DirAssetReference::new(
                     get_traced_project_dir().await?,
                     Pattern::new(pat),
-                    make_issue_source(),
+                    get_issue_source(),
                 )
                 .to_resolved()
                 .await?,
@@ -2121,7 +2131,7 @@ async fn handle_call<G: Fn(Vec<Effect>) + Send + Sync>(
                 DirAssetReference::new(
                     get_traced_project_dir().await?,
                     Pattern::new(pat),
-                    make_issue_source(),
+                    get_issue_source(),
                 )
                 .to_resolved()
                 .await?,
@@ -2412,7 +2422,7 @@ async fn handle_call<G: Fn(Vec<Effect>) + Send + Sync>(
                                 DirAssetReference::new(
                                     get_traced_project_dir().await?,
                                     Pattern::new(abs_pattern),
-                                    make_issue_source(),
+                                    get_issue_source(),
                                 )
                                 .to_resolved()
                                 .await?,
@@ -2479,7 +2489,7 @@ async fn handle_call<G: Fn(Vec<Effect>) + Send + Sync>(
                     DirAssetReference::new(
                         get_traced_project_dir().await?,
                         Pattern::new(abs_pattern),
-                        make_issue_source(),
+                        get_issue_source(),
                     )
                     .to_resolved()
                     .await?,
@@ -2548,7 +2558,7 @@ async fn handle_call<G: Fn(Vec<Effect>) + Send + Sync>(
                         DirAssetReference::new(
                             context_dir.clone(),
                             Pattern::new(Pattern::Constant(dir.into())),
-                            make_issue_source(),
+                            get_issue_source(),
                         )
                         .to_resolved()
                     })

--- a/turbopack/crates/turbopack-ecmascript/src/references/mod.rs
+++ b/turbopack/crates/turbopack-ecmascript/src/references/mod.rs
@@ -1726,6 +1726,42 @@ async fn handle_call<G: Fn(Vec<Effect>) + Send + Sync>(
                 // Ignore (e.g. dynamic parameter or string literal), just as Webpack does
                 return Ok(());
             }
+            JsValue::WellKnownFunction(WellKnownFunctionKind::NodeWorkerConstructor)
+                if analysis.analyze_mode.is_tracing() =>
+            {
+                // Only for tracing, not for bundling
+                let args = linked_args(args).await?;
+                if let Some(filename) = args.first() {
+                    let pat = js_value_to_pattern(filename);
+                    if !pat.has_constant_parts() {
+                        let (args, hints) = explain_args(&args);
+                        handler.span_warn_with_code(
+                            span,
+                            &format!("new Worker({args}) is very dynamic{hints}",),
+                            DiagnosticId::Lint(
+                                errors::failed_to_analyze::ecmascript::NEW_WORKER.to_string(),
+                            ),
+                        );
+                        if ignore_dynamic_requests {
+                            return Ok(());
+                        }
+                    }
+
+                    analysis.add_reference_code_gen(
+                        WorkerAssetReference::new(
+                            origin,
+                            Request::parse(pat).to_resolved().await?,
+                            issue_source(source, span),
+                            in_try,
+                        ),
+                        ast_path.to_vec().into(),
+                    );
+
+                    return Ok(());
+                }
+                // Ignore (e.g. dynamic parameter or string literal), just as Webpack does
+                return Ok(());
+            }
             _ => {}
         }
 
@@ -3147,7 +3183,7 @@ async fn value_visitor_inner(
             ),
             "define" => JsValue::WellKnownFunction(WellKnownFunctionKind::Define),
             "URL" => JsValue::WellKnownFunction(WellKnownFunctionKind::URLConstructor),
-            "process" => JsValue::WellKnownObject(WellKnownObjectKind::NodeProcess),
+            "process" => JsValue::WellKnownObject(WellKnownObjectKind::NodeProcessModule),
             "Object" => JsValue::WellKnownObject(WellKnownObjectKind::GlobalObject),
             "Buffer" => JsValue::WellKnownObject(WellKnownObjectKind::NodeBuffer),
             _ => return Ok((v, false)),

--- a/turbopack/crates/turbopack-ecmascript/src/references/node.rs
+++ b/turbopack/crates/turbopack-ecmascript/src/references/node.rs
@@ -1,11 +1,20 @@
 use anyhow::Result;
+use tracing::Instrument;
 use turbo_rcstr::RcStr;
 use turbo_tasks::{ResolvedVc, ValueToString, Vc};
 use turbo_tasks_fs::FileSystemPath;
 use turbopack_core::{
-    file_source::FileSource, raw_module::RawModule, reference::ModuleReference,
-    resolve::ModuleResolveResult,
+    chunk::{ChunkableModuleReference, ChunkingType, ChunkingTypeOption},
+    context::AssetContext,
+    file_source::FileSource,
+    issue::IssueSource,
+    raw_module::RawModule,
+    reference::ModuleReference,
+    reference_type::{ReferenceType, WorkerReferenceSubType},
+    resolve::{ModuleResolveResult, pattern::Pattern, resolve_raw},
 };
+
+use crate::references::util::check_and_emit_too_many_matches_warning;
 
 #[turbo_tasks::value]
 #[derive(Hash, Clone, Debug)]
@@ -43,6 +52,92 @@ impl ValueToString for PackageJsonReference {
                 self.package_json.value_to_string().await?
             )
             .into(),
+        ))
+    }
+}
+
+#[turbo_tasks::value]
+#[derive(Hash, Debug)]
+pub struct FilePathModuleReference {
+    asset_context: ResolvedVc<Box<dyn AssetContext>>,
+    context_dir: FileSystemPath,
+    path: ResolvedVc<Pattern>,
+    collect_affecting_sources: bool,
+    issue_source: IssueSource,
+}
+
+#[turbo_tasks::value_impl]
+impl FilePathModuleReference {
+    #[turbo_tasks::function]
+    pub fn new(
+        asset_context: ResolvedVc<Box<dyn AssetContext>>,
+        context_dir: FileSystemPath,
+        path: ResolvedVc<Pattern>,
+        collect_affecting_sources: bool,
+        issue_source: IssueSource,
+    ) -> Vc<Self> {
+        Self {
+            asset_context,
+            context_dir,
+            path,
+            collect_affecting_sources,
+            issue_source,
+        }
+        .cell()
+    }
+}
+
+// A reference to an module by absolute or cwd-relative file path (e.g. for the
+// worker-threads `new Worker` which has the resolving behavior of `fs.readFile` but should treat
+// the resolve result as an module instead of a raw source).
+#[turbo_tasks::value_impl]
+impl ModuleReference for FilePathModuleReference {
+    #[turbo_tasks::function]
+    async fn resolve_reference(&self) -> Result<Vc<ModuleResolveResult>> {
+        let span = tracing::info_span!(
+            "trace module",
+            pattern = display(self.path.to_string().await?)
+        );
+        async {
+            let result = resolve_raw(
+                self.context_dir.clone(),
+                *self.path,
+                self.collect_affecting_sources,
+                /* force_in_lookup_dir */ false,
+            );
+            let result = self.asset_context.process_resolve_result(
+                result,
+                ReferenceType::Worker(WorkerReferenceSubType::NodeWorker),
+            );
+
+            check_and_emit_too_many_matches_warning(
+                result,
+                self.issue_source,
+                self.context_dir.clone(),
+                self.path,
+            )
+            .await?;
+
+            Ok(result)
+        }
+        .instrument(span)
+        .await
+    }
+}
+#[turbo_tasks::value_impl]
+impl ChunkableModuleReference for FilePathModuleReference {
+    #[turbo_tasks::function]
+    fn chunking_type(&self) -> Vc<ChunkingTypeOption> {
+        Vc::cell(Some(ChunkingType::Traced))
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl ValueToString for FilePathModuleReference {
+    #[turbo_tasks::function]
+    async fn to_string(&self) -> Result<Vc<RcStr>> {
+        Ok(Vc::cell(
+            format!("raw asset {}", self.path.to_string().await?,).into(),
         ))
     }
 }

--- a/turbopack/crates/turbopack-ecmascript/src/references/raw.rs
+++ b/turbopack/crates/turbopack-ecmascript/src/references/raw.rs
@@ -6,10 +6,7 @@ use turbo_tasks_fs::FileSystemPath;
 use turbopack_core::{
     chunk::{ChunkableModuleReference, ChunkingType, ChunkingTypeOption},
     file_source::FileSource,
-    issue::{
-        Issue, IssueExt, IssueSeverity, IssueSource, IssueStage, OptionIssueSource,
-        OptionStyledString, StyledString,
-    },
+    issue::IssueSource,
     raw_module::RawModule,
     reference::ModuleReference,
     resolve::{
@@ -18,6 +15,9 @@ use turbopack_core::{
         resolve_raw,
     },
 };
+
+use crate::references::util::check_and_emit_too_many_matches_warning;
+
 #[turbo_tasks::value]
 #[derive(Hash, Debug)]
 pub struct FileSourceReference {
@@ -233,82 +233,5 @@ impl ValueToString for DirAssetReference {
         Ok(Vc::cell(
             format!("directory assets {}", self.path.to_string().await?,).into(),
         ))
-    }
-}
-
-/// If a pattern resolves to more than 10000 results, it's likely a mistake so issue a warning.
-const TOO_MANY_MATCHES_LIMIT: usize = 10000;
-
-async fn check_and_emit_too_many_matches_warning(
-    result: Vc<ModuleResolveResult>,
-    issue_source: IssueSource,
-    context_dir: FileSystemPath,
-    pattern: ResolvedVc<Pattern>,
-) -> Result<()> {
-    let num_matches = result.await?.primary.len();
-    if num_matches > TOO_MANY_MATCHES_LIMIT {
-        TooManyMatchesWarning {
-            source: issue_source,
-            context_dir,
-            num_matches,
-            pattern,
-        }
-        .resolved_cell()
-        .emit();
-    }
-    Ok(())
-}
-
-#[turbo_tasks::value(shared)]
-struct TooManyMatchesWarning {
-    source: IssueSource,
-    context_dir: FileSystemPath,
-    num_matches: usize,
-    pattern: ResolvedVc<Pattern>,
-}
-
-#[turbo_tasks::value_impl]
-impl Issue for TooManyMatchesWarning {
-    #[turbo_tasks::function]
-    async fn title(&self) -> Result<Vc<StyledString>> {
-        Ok(StyledString::Text(
-            format!(
-                "The file pattern {pattern} matches {num_matches} files in {context_dir}",
-                pattern = self.pattern.to_string().await?,
-                context_dir = self.context_dir.value_to_string().await?,
-                num_matches = self.num_matches
-            )
-            .into(),
-        )
-        .cell())
-    }
-
-    #[turbo_tasks::function]
-    fn description(&self) -> Vc<OptionStyledString> {
-        Vc::cell(Some(
-            StyledString::Text(rcstr!(
-                "Overly broad patterns can lead to build performance issues and over bundling."
-            ))
-            .resolved_cell(),
-        ))
-    }
-
-    #[turbo_tasks::function]
-    async fn file_path(&self) -> Vc<FileSystemPath> {
-        self.source.file_path()
-    }
-
-    #[turbo_tasks::function]
-    fn stage(&self) -> Vc<IssueStage> {
-        IssueStage::Resolve.cell()
-    }
-
-    fn severity(&self) -> IssueSeverity {
-        IssueSeverity::Warning
-    }
-
-    #[turbo_tasks::function]
-    fn source(&self) -> Vc<OptionIssueSource> {
-        Vc::cell(Some(self.source))
     }
 }

--- a/turbopack/crates/turbopack-ecmascript/src/references/util.rs
+++ b/turbopack/crates/turbopack-ecmascript/src/references/util.rs
@@ -1,8 +1,16 @@
 use anyhow::Result;
 use swc_core::{ecma::ast::Expr, quote};
 use turbo_rcstr::{RcStr, rcstr};
-use turbo_tasks::Vc;
-use turbopack_core::{self, resolve::parse::Request};
+use turbo_tasks::{ResolvedVc, ValueToString, Vc};
+use turbo_tasks_fs::FileSystemPath;
+use turbopack_core::{
+    self,
+    issue::{
+        Issue, IssueExt, IssueSeverity, IssueSource, IssueStage, OptionIssueSource,
+        OptionStyledString, StyledString,
+    },
+    resolve::{ModuleResolveResult, parse::Request, pattern::Pattern},
+};
 
 /// Creates a IIFE expression that throws a "Cannot find module" error for the
 /// given request string
@@ -35,4 +43,81 @@ pub async fn request_to_string(request: Vc<Request>) -> Result<Vc<RcStr>> {
             // TODO: Handle Request::Dynamic, Request::Alternatives
             .unwrap_or(rcstr!("unknown")),
     ))
+}
+
+/// If a pattern resolves to more than 10000 results, it's likely a mistake so issue a warning.
+const TOO_MANY_MATCHES_LIMIT: usize = 10000;
+
+pub async fn check_and_emit_too_many_matches_warning(
+    result: Vc<ModuleResolveResult>,
+    issue_source: IssueSource,
+    context_dir: FileSystemPath,
+    pattern: ResolvedVc<Pattern>,
+) -> Result<()> {
+    let num_matches = result.await?.primary.len();
+    if num_matches > TOO_MANY_MATCHES_LIMIT {
+        TooManyMatchesWarning {
+            source: issue_source,
+            context_dir,
+            num_matches,
+            pattern,
+        }
+        .resolved_cell()
+        .emit();
+    }
+    Ok(())
+}
+
+#[turbo_tasks::value(shared)]
+struct TooManyMatchesWarning {
+    source: IssueSource,
+    context_dir: FileSystemPath,
+    num_matches: usize,
+    pattern: ResolvedVc<Pattern>,
+}
+
+#[turbo_tasks::value_impl]
+impl Issue for TooManyMatchesWarning {
+    #[turbo_tasks::function]
+    async fn title(&self) -> Result<Vc<StyledString>> {
+        Ok(StyledString::Text(
+            format!(
+                "The file pattern {pattern} matches {num_matches} files in {context_dir}",
+                pattern = self.pattern.to_string().await?,
+                context_dir = self.context_dir.value_to_string().await?,
+                num_matches = self.num_matches
+            )
+            .into(),
+        )
+        .cell())
+    }
+
+    #[turbo_tasks::function]
+    fn description(&self) -> Vc<OptionStyledString> {
+        Vc::cell(Some(
+            StyledString::Text(rcstr!(
+                "Overly broad patterns can lead to build performance issues and over bundling."
+            ))
+            .resolved_cell(),
+        ))
+    }
+
+    #[turbo_tasks::function]
+    async fn file_path(&self) -> Vc<FileSystemPath> {
+        self.source.file_path()
+    }
+
+    #[turbo_tasks::function]
+    fn stage(&self) -> Vc<IssueStage> {
+        IssueStage::Resolve.cell()
+    }
+
+    fn severity(&self) -> IssueSeverity {
+        IssueSeverity::Warning
+    }
+
+    #[turbo_tasks::function]
+    fn source(&self) -> Vc<OptionIssueSource> {
+        Vc::cell(Some(self.source))
+    }
 }

--- a/turbopack/crates/turbopack-ecmascript/src/static_code.rs
+++ b/turbopack/crates/turbopack-ecmascript/src/static_code.rs
@@ -16,7 +16,6 @@ use crate::EcmascriptAnalyzable;
 /// the final runtime code, while keeping source map information.
 #[turbo_tasks::value]
 pub struct StaticEcmascriptCode {
-    asset_context: ResolvedVc<Box<dyn AssetContext>>,
     asset: ResolvedVc<Box<dyn EcmascriptAnalyzable>>,
     generate_source_map: bool,
 }
@@ -42,7 +41,6 @@ impl StaticEcmascriptCode {
             bail!("asset is not an Ecmascript module")
         };
         Ok(Self::cell(StaticEcmascriptCode {
-            asset_context,
             asset,
             generate_source_map,
         }))

--- a/turbopack/crates/turbopack-ecmascript/src/utils.rs
+++ b/turbopack/crates/turbopack-ecmascript/src/utils.rs
@@ -197,12 +197,17 @@ pub fn module_value_to_well_known_object(module_value: &ModuleValue) -> Option<J
         }
         "node:fs" | "fs" => JsValue::WellKnownObject(WellKnownObjectKind::FsModule),
         "node:child_process" | "child_process" => {
-            JsValue::WellKnownObject(WellKnownObjectKind::ChildProcess)
+            JsValue::WellKnownObject(WellKnownObjectKind::ChildProcessModule)
         }
         "node:os" | "os" => JsValue::WellKnownObject(WellKnownObjectKind::OsModule),
-        "node:process" | "process" => JsValue::WellKnownObject(WellKnownObjectKind::NodeProcess),
+        "node:process" | "process" => {
+            JsValue::WellKnownObject(WellKnownObjectKind::NodeProcessModule)
+        }
         "node:url" | "url" => JsValue::WellKnownObject(WellKnownObjectKind::UrlModule),
         "node:module" | "module" => JsValue::WellKnownObject(WellKnownObjectKind::ModuleModule),
+        "node:worker_threads" | "worker_threads" => {
+            JsValue::WellKnownObject(WellKnownObjectKind::WorkerThreadsModule)
+        }
         "node-pre-gyp" | "@mapbox/node-pre-gyp" => {
             JsValue::WellKnownObject(WellKnownObjectKind::NodePreGyp)
         }

--- a/turbopack/crates/turbopack-nft/src/nft.rs
+++ b/turbopack/crates/turbopack-nft/src/nft.rs
@@ -76,7 +76,7 @@ async fn node_file_trace_operation(
     let environment = Environment::new(ExecutionEnvironment::NodeJsLambda(
         NodeJsEnvironment::default().resolved_cell(),
     ));
-    let module_asset_context = ModuleAssetContext::new(
+    let module_asset_context = ModuleAssetContext::new_without_replace_externals(
         Default::default(),
         // This config should be kept in sync with
         // turbopack/crates/turbopack-tracing/tests/node-file-trace.rs and
@@ -109,6 +109,7 @@ async fn node_file_trace_operation(
             enable_node_native_modules: true,
             enable_node_modules: Some(input_dir),
             custom_conditions: vec![rcstr!("node")],
+            enable_node_externals: true,
             loose_errors: true,
             collect_affecting_sources: true,
             ..Default::default()

--- a/turbopack/crates/turbopack-tracing/tests/node-file-trace.rs
+++ b/turbopack/crates/turbopack-tracing/tests/node-file-trace.rs
@@ -358,7 +358,7 @@ async fn node_file_trace_operation(
         }
         .resolved_cell(),
     ));
-    let module_asset_context = ModuleAssetContext::new(
+    let module_asset_context = ModuleAssetContext::new_without_replace_externals(
         Default::default(),
         // TODO These test cases should move into the `node-file-trace` crate and use the same
         // config.

--- a/turbopack/crates/turbopack-tracing/tests/node-file-trace.rs
+++ b/turbopack/crates/turbopack-tracing/tests/node-file-trace.rs
@@ -149,6 +149,7 @@ static ALLOC: turbo_tasks_malloc::TurboMalloc = turbo_tasks_malloc::TurboMalloc;
 #[case::pdf2json("integration/pdf2json.mjs")]
 #[case::pdfkit("integration/pdfkit.js")]
 #[case::pg("integration/pg.js")]
+#[case::pino("integration/pino.js")]
 #[case::playwright_core("integration/playwright-core.js")]
 #[case::pnpm_like("integration/pnpm/pnpm-like.js")]
 #[case::polyfill_library("integration/polyfill-library.js")]

--- a/turbopack/crates/turbopack-tracing/tests/node-file-trace/integration/pino.js
+++ b/turbopack/crates/turbopack-tracing/tests/node-file-trace/integration/pino.js
@@ -1,0 +1,18 @@
+// for transport({target: 'pino-pretty'}) to work
+require('pino-pretty')
+
+const { pino, transport } = require('pino')
+
+const loggerOptions = {
+  level: 'info',
+  timestamp: false,
+}
+const loggerTransport = transport({
+  target: 'pino-pretty',
+  options: {
+    ignore: 'pid,hostname',
+  },
+})
+const logger = pino(loggerOptions, loggerTransport)
+
+logger.info('foo')

--- a/turbopack/crates/turbopack-tracing/tests/node-file-trace/package.json
+++ b/turbopack/crates/turbopack-tracing/tests/node-file-trace/package.json
@@ -86,6 +86,8 @@
     "pg": "^7.11.0",
     "phantomjs-prebuilt": "^2.1.16",
     "pixelmatch": "^5.3.0",
+    "pino": "^10.1.0",
+    "pino-pretty": "^13.1.2",
     "playwright-core": "^1.17.1",
     "polyfill-library": "3.93.0",
     "prismjs": "^1.25.0",

--- a/turbopack/crates/turbopack-tracing/tests/unit.rs
+++ b/turbopack/crates/turbopack-tracing/tests/unit.rs
@@ -194,7 +194,7 @@ async fn node_file_trace_operation(package_root: RcStr, input: RcStr) -> Result<
         }
         .resolved_cell(),
     ));
-    let module_asset_context = ModuleAssetContext::new(
+    let module_asset_context = ModuleAssetContext::new_without_replace_externals(
         Default::default(),
         // TODO These test cases should move into the `node-file-trace` crate and use the same
         // config.

--- a/turbopack/crates/turbopack/src/lib.rs
+++ b/turbopack/crates/turbopack/src/lib.rs
@@ -391,8 +391,9 @@ impl ModuleAssetContext {
         })
     }
 
+    /// Doesn't replace external resolve results with a CachedExternalModule.
     #[turbo_tasks::function]
-    fn new_without_replace_externals(
+    pub fn new_without_replace_externals(
         transitions: ResolvedVc<TransitionOptions>,
         compile_time_info: ResolvedVc<CompileTimeInfo>,
         module_options_context: ResolvedVc<ModuleOptionsContext>,

